src/mozanalysis/metrics/__init__.py

```diff
@@ -2,345 +2,118 @@
-                this ``Metric``.
-
-        The wrapped function needs to have one argument (the DataFrame
-        output by ``data_source.get_dataframe()``), and return a Column.
-        The function's name is taken as the ``Metric``'s name.
+    data_source = attr.ib(type=DataSource)
+    select_expr = attr.ib(type=str)
 
-        Example::
 
-            clients_daily = DataSource.from_table_name('clients_daily')
-
-            @Metric.from_func(clients_daily)
-            def active_hours(cd: DataFrame) -> Column:
-                return agg_sum(cd.active_hours_sum)
-        """
-        def f(col_getter):
-            return cls(
-                name=col_getter.__name__,
-                data_source=data_source,
-                col_getter=col_getter,
-                description=col_getter.__doc__
-            )
-
-        return f
-
-    @classmethod
-    def from_col(cls, metric_name, col, data_source):
-        """Return a ``Metric`` defined by ``col``.
-
-        Args:
-            metric_name (str): Name for the resulting Column.
-            col (Column): Spark Column representing the computed metric.
-            data_source (DataSource): DataSource wrapper for the source
-                DataFrame for the Column.
-
-        Example::
-
-            cd_ds = DataSource.from_table_name('clients_daily')
-
-            active_hours = Metric.from_col(
-                name='active_hours',
-                col=agg_sum(cd.active_hours_sum),
-                data_source=cd_ds
-            )
-        """
-        return cls(
-            name=metric_name,
-            data_source=data_source,
-            col_getter=lambda _: col
-        )
-
-    def get_col(self, spark, experiment):
-        """Return a ``Column`` representing this metric.
-
-        Args:
-            spark (SparkSession): The Spark Session.
-            experiment (Experiment): The experiment for which this
-                metric is being computed.
-        """
-        # TODO: will this ever need access to ``enrollments``?
-        df = self.data_source.get_dataframe(spark, experiment)
-
-        try:
-            col = self._col_getter(df)
-        except TypeError:
-            col = self._col_getter(df, experiment)
-        return col.alias(self.name)
-
-
-def agg_sum(col):
-    """Return the sum over the data, with 0-filled nulls.
-
-    Args:
-        col (Column): A Spark Column of a numeric type
+def agg_sum(select_expr):
+    """Return a SQL fragment for the sum over the data, with 0-filled nulls.
     """
-    return F.coalesce(F.sum(col), F.lit(0))
+    return "COALESCE(SUM({}), 0)".format(select_expr)
 
 
-def agg_any(col):
-    """Return the bitwise OR, as an int, with 0-filled nulls.
-
-    Args:
-        col (Column): A Spark Column of bools
-    """
-    return F.coalesce(F.max(col).astype('int'), F.lit(0))
+def agg_any(select_expr):
+    """Return the bitwise OR, as an int, with 0-filled nulls."""
```

Comment:
s/bitwise/logical/

(Why do we prefer an integer here?)

Reply:
Whoa! This is Spark code. Can you tell that we don't unit test our metrics?

At least colab's `!pip install` can install branches from github, so I'll try out these metrics before merging (after fixing this code)

Reply:
I think it's correctly operating elementwise -- it's just not doing bitwise operations (e.g. the result of "`agg_any([1, 64])`" will be 1, not 65)

Reply:
> Whoa! This is Spark code

:man_facepalming: This is _removing_ Spark code. I was not in good shape yesterday!

> s/bitwise/logical/

[lol](https://github.com/mozilla/mozanalysis/pull/26/files/a13e18c20d7021daf6e35a90ca9e19240472d641#diff-71962d4e0d4376ffe13d532605fbfbf8)

> Why do we prefer an integer here

Good question; this is a hangover from our spark datasets. I'll have to check whether we can now rely on things being boolean.

Reply:
I've fixed this; there is one remaining question in my mind: do we stick with "any/all" naming or do we switch to "or/and" naming (i.e. rename it `agg_or()`).

IIRC, in spark land, "any/all" was a no-brainer because it was consistent with spark. But we're not there any more.

I'm leaning towards keeping `agg_any()` because I think it's clearer in the context in which `agg_any()` is being used:
```
unenroll = Metric(
    name='unenroll',
    data_source=events,
    select_expr=agg_any("""
            event_category = 'normandy'
            AND event_method = 'unenroll'
            AND event_string_value = '{experiment_slug}'
        """)
)
```